incubator/hnc/internal/pkg/selector/selector.go

```diff
@@ -0,0 +1,67 @@
+package selector
+
+import (
+	"strings"
+
+	"github.com/go-logr/logr"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+
+	api "sigs.k8s.io/multi-tenancy/incubator/hnc/api/v1alpha2"
+)
+
+func GetSelectorAnnotation(inst *unstructured.Unstructured) (string, bool) {
+	annot := inst.GetAnnotations()
+	str, ok := annot[api.AnnotationSelector]
+	return str, ok
+}
+
+// GetTreeSelector is similar to a regular selector, except that it adds the LabelTreeDepthSuffix to every string
+// To transform a tree selector into a regular label selector, we follow these steps:
+// 1. get the treeSelector annotation if it exists
+// 2. convert the annotation string to a slice of strings seperated by comma, because user is allowed to put multiple selectors
+// 3. append the LabelTreeDepthSuffix to each of the treeSelector string
+// 4. combine them into a single string connected by comma
+func GetTreeSelector(log logr.Logger, inst *unstructured.Unstructured) labels.Selector {
+	annot := inst.GetAnnotations()
+	treeSelectorStr, ok := annot[api.AnnotationTreeSelector]
+	if !ok {
+		return nil
+	}
+	strs := strings.Split(treeSelectorStr, ",")
+	selectorStr := ""
+	for i, str := range strs {
+		selectorStr = selectorStr + str + api.LabelTreeDepthSuffix
+		if i < len(strs)-1 {
+			selectorStr = selectorStr + ", "
+		}
+	}
+	return GetSelectorFromString(log, selectorStr)
+}
+
+// GetSelector returns the selector on a given object if it exists
+func GetSelector(log logr.Logger, inst *unstructured.Unstructured) labels.Selector {
+	selectorStr, ok := GetSelectorAnnotation(inst)
+	if !ok {
+		return nil
+	}
+	return GetSelectorFromString(log, selectorStr)
+}
+
+// getSelectorFromString converts the given string to a selector
+// Note: any invalid Selector value will cause this object not propagating to any child namespace
+func GetSelectorFromString(log logr.Logger, str string) labels.Selector {
+	labelSelector, err := v1.ParseToLabelSelector(str)
+	// TODO: surface the error messages here (https://github.com/kubernetes-sigs/multi-tenancy/issues/1165)
+	if err != nil {
```

Comment:
We're throwing away valuable information from the label parser here by not returning the error - check to see if there's anything in there that would be useful to present to the user.

Reply:
Does `log.Error(err, "Could not parse selector annotation to labelSelector")` not log the `err`?

Reply:
done

Reply:
done